docs/contributing/coding-guidelines/ros-nodes/directory-structure.md

```diff
@@ -1,75 +1,211 @@
-│   └─ foo.launch.py
-├─ schema
-│   └─ foo_node.schema.json
+│
+├─ include  # for exporting headers
+│   └─ autoware
+│       └─ gnss_poser
+│           └─ exported_header.hpp
+│
 ├─ src
-│   ├─ foo_node.cpp
-│   ├─ foo_node.hpp
-│   └─ foo_private.hpp
-├─ test
-│   └─ test_foo.cpp
+│   ├─ include
+│   │   ├─ gnss_poser_node.hpp
+│   │   └─ foo.hpp
+│   ├─ gnss_poser_node.cpp
+│   └─ bar.cpp
+│
+├─ launch
+│   ├─ gnss_poser.launch.xml
+│   └─ gnss_poser.launch.py
+│
+└─ test
+    ├─ test_foo.hpp  # or place under an `include` folder here
+    └─ test_foo.cpp
+```
+
+### Package name
+
+- All the packages in Autoware should be prefixed with `autoware_`.
+- Even if the package is exports a node, the package name **should NOT** have the `_node` suffix.
+- The package name should be in `snake_case`.
+
+| Package Name                      | OK  | Alternative                  |
+| --------------------------------- | --- | ---------------------------- |
+| path_smoother                     | ❌  | autoware_path_smoother       |
+| autoware_trajectory_follower_node | ❌  | autoware_trajectory_follower |
+| autoware_geography_utils          | ✅  | -                            |
+
+### Package folder
+
+```txt
+autoware_gnss_poser
 ├─ package.xml
 ├─ CMakeLists.txt
 └─ README.md
 ```
 
-### Directory descriptions
+The package folder name should be the same as the package name.
 
-#### `config`
+#### `package.xml`
 
-Place configuration files such as node parameters.
-For ROS parameters, use the extension `.param.yaml`.
-For non-ROS parameters, use the extension `.yaml`.
+- The package name should be entered within the `<name>` tag.
+  - `<name>autoware_gnss_poser</name>`
 
-Rationale: Since ROS parameters files are type-sensitive, they should not be the target of some code formatters and linters. In order to distinguish the file type, we use different file extensions.
+#### `CMakeLists.txt`
 
-#### `doc`
+- The [`project()`](https://cmake.org/cmake/help/latest/command/project.html) command should call the package name.
+  - **Example:** `project(autoware_gnss_poser)`
 
-Place document files and link from README.
+##### Exporting a component
 
-#### `include`
+```cmake
+ament_auto_add_library(${PROJECT_NAME} SHARED
+  src/gnss_poser_node.cpp
+)
 
-Place header files exposed to other packages. Do not place files directly under the `include` directory, but place files under the directory with the package name.
-This directory is used for mostly library headers. Note that many headers do not need to be placed here. It is enough to place the headers under the `src` directory.
+rclcpp_components_register_node(${PROJECT_NAME}
+  PLUGIN "autoware::gnss_poser::GNSSPoserNode"
+  EXECUTABLE ${PROJECT_NAME}_node
+)
+```
 
-Reference: <https://docs.ros.org/en/humble/How-To-Guides/Ament-CMake-Documentation.html#adding-targets>
+- The component executable should have `_node` suffix.
+- The component executable name should start with `${PROJECT_NAME}`
```

Comment:
I think this is verbosity because when specifying a node, the package name functions as a namespace. Also, in some cases, there are executables that are not nodes, so it would be simpler if developers could name them freely.

Reply:
There are multiple ways devs name their executables:
- `${PROJECT_NAME}`
- `${PROJECT_NAME}_exe`
- `${PROJECT_NAME}_node`
- `${PROJECT_NAME}_node_exe`

This is for nodes, not libraries by the way.

Wouldn't it be better if the node executables were named in a set way?

Reply:
> there are executables that are not nodes

@isamu-takagi could you give examples?

Reply:
> Wouldn't it be better if the node executables were named in a set way?

Ah, surely. That should be standardized.

>>there are executables that are not nodes
>
>@isamu-takagi could you give examples?

For example, I think there are several nodes that have tools for data conversion or pre-processing. These are CLI programs.

Reply:
I see, my rules are targeting specifically `The component executable`s which are composable node components.

These lines are under `##### Exporting a component` title. Do you think I should clarify it more or should I keep them as is?

Reply:
`component executable` is a bit ambiguous, so it might be better to write it as `node` or `node executable` (only node can be executable, and composable node is exported as a class name string)

Reply:
@isamu-takagi I tried to take your concerns into account with: https://github.com/autowarefoundation/autoware-documentation/pull/547/commits/e38f3a2d556b3d4fa5353bff18ff6b309290abac

But if you don't like it, could you add a commit to apply your suggestions, your version?

Reply:
@xmfcx  Sorry for your confusion. I was confusing composable node class and composable node executable. So I reverted your commit with changing some words. Could you please check it? https://github.com/autowarefoundation/autoware-documentation/pull/547/commits/d96bd27b2552042c4ff63ca17a5532ac29d132de

What about `${PROJECT_NAME}` prefix? It is suitable if the package has only one node, but if it has multiple nodes, I would like to name them individually without a prefix (e.g. [topic_tools](https://github.com/ros-tooling/topic_tools/blob/rolling/topic_tools/CMakeLists.txt)).

Reply:
It's alright, I've brought the non-composable executable back, **while discouraging it** with: https://github.com/autowarefoundation/autoware-documentation/pull/547/commits/9ba71b95d580d47e711653315fd43c8616161241

Because some offline/non-runtime nodes may still use non-composable nodes.

And added information for handling multiple node cases like topic_tools.

Could you check again?

Reply:
Looks good. thank you!

Reply:
![kono-suba-aqua-smile](https://github.com/autowarefoundation/autoware-documentation/assets/10751153/6a80beba-0045-47f8-8ba7-30bd0189b7e2)